### DISCUSSION TITLE
Manual file upload in feature test fails with `Undefined property: Illuminate\Http\UploadedFile::$name`

### DISCUSSION THREAD

**Discussion by tedsecretsource** [livewire/livewire#3092] [7. Bugs (archived)] [answered]
## Introduction

This is a greatly reduced summary as the actual code I'm using is much more complex but I have tested the exact steps shown here and this is reproducible. As I am new to Livewire, it is _very possible_ (probable 🤔) that I'm missing something obvious.

### Description

Feature tests with real file uploads fail unless you manually set the `$file->name` property. If you don't set it, you'll get an ErrorException "Undefined property: Illuminate\Http\UploadedFile::$name".

```php
  at vendor/livewire/livewire/src/Testing/Concerns/MakesCallsToComponent.php:93
     89▕         $this->sendMessage('callMethod', [
     90▕             'method' => 'startUpload',
     91▕             'params' => [$name, collect($files)->map(function ($file) {
     92▕                 return [
  ➜  93▕                     'name' => $file->name,
     94▕                     'size' => $file->getSize(),
     95▕                     'type' => $file->getMimeType(),
     96▕                 ];
     97▕             })->toArray(), $isMultiple],
```

Given I have an application that needs to process the contents of a file upload, e.g.: rows in a spreadsheet, I cannot use `UploadedFile::fake()` because the file contents is empty. The workaround is to use an actual test file with known content and manually "upload" the file by creating a new UploadedFile object.

### Exact steps to reproduce

1. Create a Livewire component with a form with a file input element and wire it to "save". We'll call our component SpreadsheetUploader.
2. Create a Feature test called UploadsSpreadsheetTest and add the test code shown below.
3. Run `php artisan test`

Expected output:
```shell
   PASS  Tests\Feature\UploadSpreadsheetTest
  ✓ saving spreadsheet uploads

  Tests:  1 passed
  Time:   3.51s
```

Actual output:
```shell
   FAIL  Tests\Feature\UploadSpreadsheetTest
  ⨯ saving spreadsheet uploads

  ---

  • Tests\Feature\UploadSpreadsheetTest > saving spreadsheet uploads
   ErrorException 

  Undefined property: Illuminate\Http\UploadedFile::$name

  at vendor/livewire/livewire/src/Testing/Concerns/MakesCallsToComponent.php:93
     89▕         $this->sendMessage('callMethod', [
     90▕             'method' => 'startUpload',
     91▕             'params' => [$name, collect($files)->map(function ($file) {
     92▕                 return [
  ➜  93▕                     'name' => $file->name,
     94▕                     'size' => $file->getSize(),
     95▕                     'type' => $file->getMimeType(),
     96▕                 ];
     97▕             })->toArray(), $isMultiple],

      +1 vendor frames 
  2   [internal]:0
      Livewire\Testing\TestableLivewire::Livewire\Testing\Concerns\{closure}()

      +5 vendor frames 
  8   tests/Feature/UploadSpreadsheetTest.php:48
      Livewire\Testing\TestableLivewire::set()


  Tests:  1 failed
  Time:   1.41s

```

### Stripped-down, copy-pastable code snippets that have been greatly simplified to improve clarity

app/Livewire/SpreadsheetUploader.php
```php
namespace App\Http\Livewire;

use Livewire\Component;
use Livewire\WithFileUploads;

class SpreadsheetUploader extends Component
{
    use WithFileUploads;

    public $members;
    public $storedPath = '';

    public function save()
    {
        $this->storedPath = $this->members->store('spreadsheets');
        /* Do stuff with the contents of the spreadsheet */
        // Save each row to the database…
    }

    public function render()
    {
        return view('livewire.spreadsheet-uploader');
    }
}

````

resources/views/livewire/spreadsheet-uploader.blade.php
```html
<div>
<form wire:model.prevent="save" method="post">
    <input wire:model="members" type="file" />
    <button type="submit">Save</button>
</form>
@if($storedPath)
<p>The spreadsheet has been uploaded</p>
@endif
</div>
```

#### This test fails with an exception
tests/Feature/UploadsSpreadsheetTest-fails.php
```php
$file = new UploadedFile(
    'tests/Feature/members_test_data.xlsx',
    'members_test_data.xlsx',
    'application/vnd.ms-excel',
    null,
    true
);

Livewire::test('emailer-dashboard')
    ->set('members', $file)
    ->call('save')
    ->assertSee('The spreadsheet has been uploaded');

/* Run additional assertions, like, see if the data has been stored in the database */
```

#### This test runs without exceptions
tests/Feature/UploadsSpreadsheetTest-succeeds.php
```php
$file = new UploadedFile(
    'tests/Feature/members_test_data.xlsx',
    'members_test_data.xlsx',
    'application/vnd.ms-excel',
    null,
    true
);

/* The line below makes the test run without exceptions */
$file->name = 'members_test_data.xlsx';

Livewire::test('spreadsheet-uploader')
    ->set('members', $file)
    ->call('save')
    ->assertSee('The spreadsheet has been uploaded');

/* Run additional assertions, like, see if the data has been stored in the database */
```

Note the line: `$file->name = 'members_test_data.xlsx';`

**Context**
- Livewire version: 2.3.5
- Laravel version: 8.20.1

**Comment by reniar1** [comment]
I am having the same problem 

  public function can_upload_video()
    {
        $path = '/Users/reniar/code/OfflineJetstream/storage/app/videos/test.mp4';
        $name = 'video';
        $file = new UploadedFile($path, $name, 'video/mp4', 4, false);

        $sizeInKilobytes = 100000;
        $file = UploadedFile::fake()->create(
            'video.mp4', $sizeInKilobytes, 'video/mp4'
        );

        Storage::fake('videos_disk');

        $this->actingAs(User::factory()->create());

        Livewire::test(Create::class)
            ->set('title', 'Testvideo333')
            ->set('video', $file)
            ->call('upload');

        dd(Storage::disk('videos_disk')->allFiles());
    }

**My guess is that something here might be wrong**

```
 public function syncUploadedFiles($name, $files, $isMultiple = false)
    {
        // This methhod simulates the calls Livewire's JavaScript
        // normally makes for file uploads.
        $this->sendMessage('callMethod', [
            'method' => 'startUpload',
            'params' => [$name, collect($files)->map(function ($file) {
                return [
                    **'name' => $file->name,**
                    'size' => $file->getSize(),
                    'type' => $file->getMimeType(),
                ];
            })->toArray(), $isMultiple],
        ]);

```

the faked object passes validation and the new UploadedFile object doenst pass the validation even if I will assign the name property manually like @tedsecretsource alredy mentioned

**Comment by RJ2020DD** [comment, marked as answer]
I had the same issue but found that using 'createWithContent()' solved it for me:

```
UploadedFile::fake()->createWithContent('spreadsheet_template.xlsx', file_get_contents(base_path( 'tests/Helpers/valid-spreadsheet.xlsx' )));
```

Just sharing in case it helps :)

**Comment by tedsecretsource** [comment]
> I had the same issue but found that using 'createWithContent()' solved it for me:
> 
> ```
> UploadedFile::fake()->createWithContent('spreadsheet_template.xlsx', file_get_contents(base_path( 'tests/Helpers/valid-spreadsheet.xlsx' )));
> ```
> 
> Just sharing in case it helps :)

Marking this as the answer although I haven't tried it because it looks exactly like what is required.
